executor/join.go

```diff
@@ -258,32 +253,49 @@ func (e *HashJoinExec) wait4Inner() (finished bool, err error) {
 			return false, err
 		}
 	}
-	if e.innerResult.Len() == 0 && (e.joinType == plannercore.InnerJoin || e.joinType == plannercore.SemiJoin) {
+	if e.hashTable.Len() == 0 && (e.joinType == plannercore.InnerJoin || e.joinType == plannercore.SemiJoin) {
 		return true, nil
 	}
 	return false, nil
 }
 
 var innerResultLabel fmt.Stringer = stringutil.StringerStr("innerResult")
 
-// fetchInnerRows fetches all rows from inner executor, and append them to
-// e.innerResult.
-func (e *HashJoinExec) fetchInnerRows(ctx context.Context) error {
-	e.innerResult = chunk.NewList(retTypes(e.innerExec), e.initCap, e.maxChunkSize)
+// fetchInnerRows fetches all rows from inner executor,
+// and append them to e.innerResult.
+func (e *HashJoinExec) fetchInnerRows(ctx context.Context, chkCh chan<- *chunk.Chunk, doneCh chan struct{}) {
+	defer close(chkCh)
+	e.innerResult = chunk.NewList(e.innerExec.base().retFieldTypes, e.initCap, e.maxChunkSize)
 	e.innerResult.GetMemTracker().AttachTo(e.memTracker)
 	e.innerResult.GetMemTracker().SetLabel(innerResultLabel)
 	var err error
 	for {
-		if e.finished.Load().(bool) {
-			return nil
-		}
-
-		chk := newFirstChunk(e.children[e.innerIdx])
-		err = Next(ctx, e.innerExec, chunk.NewRecordBatch(chk))
-		if err != nil || chk.NumRows() == 0 {
-			return err
+		select {
+		case <-doneCh:
+			return
+		case <-e.closeCh:
+			return
+		default:
+			if e.finished.Load().(bool) {
+				return
+			}
+			chk := newFirstChunk(e.children[e.innerIdx])
```

Comment:
`newFirstChunk()` allocates a Chunk with `tidb_init_chunk` size. I think we'd better allocate `tidb_max_chunk_size` rows for each `Chunk` used to store inner rows. Because building hash table is a pipeline breaker.